module_checksum_upgrade/models/module.py

```diff
@@ -0,0 +1,134 @@
+        Icp = self.env['ir.config_parameter']
+        Icp.set_param(PARAM_INSTALLED_CHECKSUMS, json.dumps(checksums))
+
+    @api.model
+    def _save_installed_checksums(self):
+        checksums = {}
+        installed_modules = self.search([('state', '=', 'installed')])
+        for module in installed_modules:
+            checksums[module.name] = module._get_checksum_dir()
+        self._save_checksums(checksums)
+
+    @api.model
+    def _get_modules_partially_installed(self):
+        return self.search([
+            '!',
+            ('state', 'in', ('installed', 'uninstalled', 'uninstallable')),
+        ])
+
+    @api.model
+    def _get_modules_with_changed_checksum(self):
+        saved_checksums = self._get_saved_checksums()
+        installed_modules = self.search([('state', '=', 'installed')])
+        return installed_modules.filtered(
+            lambda r: r._get_checksum_dir() != saved_checksums.get(r.name),
+        )
+
+    @api.model
+    def upgrade_changed_checksum(self):
+        """ Run an upgrade of the database, upgrading only changed modules.
+
+        Installed modules for which the checksum has changed since the
+        last successful run of this method are marked "to upgrade",
+        then the normal Odoo scheduled upgrade process
+        is launched.
+
+        If there is no module with a changed checksum, and no module in state
+        other than installed, uninstalled, uninstallable, this method does
+        nothing, otherwise the normal Odoo upgrade process is launched.
+
+        After a successful upgrade, the checksums of installed modules are
+        saved.
+
+        In case of error during the upgrade, an exception is raised.
+        If any module remains to upgrade or to uninstall after the upgrade
+        process, this method returns False. It returns True only in case
+        of complete success.
+        """
+        _logger.info("Upgrade starting...")
+
+        _logger.info("Updating modules list...")
+        self.update_list()
+        changed_modules = self._get_modules_with_changed_checksum()
+        if not changed_modules and not self._get_modules_partially_installed():
+            _logger.info("No checksum change detected in installed modules "
+                         "and all modules installed, nothing to do.")
+            return True
+        _logger.info("Marking the following module to upgrade "
+                     "following checksum change: %s...",
+                     ','.join(changed_modules.mapped('name')))
+        changed_modules.button_upgrade()
+        self.env.cr.commit()  # pylint: disable=invalid-commit
+
+        _logger.info("Upgrading...")
+        self.env['base.module.upgrade'].upgrade_module()
+        self.env.cr.commit()  # pylint: disable=invalid-commit
+
+        _logger.info("Upgrade successful, updating checksums...")
+        self._save_installed_checksums()
+        self.env.cr.commit()  # pylint: disable=invalid-commit
```

Comment:
Could you explain why the commits instead of savepoints? I'm pretty sure it makes sense, but a comment would be quite useful. :wink:

Reply:
What would be the benefit of savepoints?

Reply:
Savepoints produce the same effect as a commit, but just for the current cursor. At the end, the cursor is committed automatically by Odoo following its usual rules.

Please read this section: https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#never-commit-the-transaction

Reply:
Savepoint are not the same as commits: https://www.postgresql.org/docs/9.6/static/sql-savepoint.html.

I'm aware of the risks of commits in normal business logic. Here I really want to commit at each step (to be sure also to have the same behaviour with each Odoo version). There is no risk of corruption because this method is not intended to be invoked has part of a larger transaction. I'll update the docstring on this topic.

Reply:
Docstring updated